presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,788 @@
+            return;
+        }
+
+        long oldSequenceId = currentSequenceId.get();
+        int pagesToRemove = toIntExact(sequenceId - oldSequenceId);
+        long currentSequenceId = oldSequenceId;
+
+        checkArgument(pagesToRemove <= totalPagesRemaining.get(), "Invalid sequenceId");
+
+        // remove the pages from storage
+        currentSequenceId += acknowledgePagesFromStorage(sequenceId);
+
+        // remove the pages from memory
+        if (currentSequenceId < sequenceId) {
+            acknowledgePagesFromMemory(sequenceId, currentSequenceId);
+        }
+
+        verify(this.currentSequenceId.compareAndSet(oldSequenceId, oldSequenceId + pagesToRemove));
+    }
+
+    private synchronized long acknowledgePagesFromStorage(long sequenceId)
+    {
+        long pagesAcknowledged = 0;
+        long pagesRemoved = 0;
+        long bytesRemoved = 0;
+        List<HandleInfo> handleInfos = ImmutableList.copyOf(handleInfoQueue);
+
+        for (HandleInfo handleInfo : handleInfos) {
+            Range<Long> range = handleInfo.getRange();
+
+            if (range.upperEndpoint() <= sequenceId) {
+                handleInfo.removeFile();
+                handleInfoQueue.removeFirst();
+
+                pagesAcknowledged += handleInfo.getPageCount() - startPage.get();
+                pagesRemoved += handleInfo.getPageCount();
+                bytesRemoved += handleInfo.getBytes();
+
+                startPage.set(0);
+            }
+            else {
+                pagesAcknowledged += sequenceId - range.lowerEndpoint() - startPage.get();
+                startPage.set(toIntExact(sequenceId - range.lowerEndpoint()));
+                break;
+            }
+        }
+
+        totalBufferedPages.addAndGet(-pagesRemoved);
+        totalBufferedBytes.addAndGet(-bytesRemoved);
+        totalPagesRemaining.addAndGet(-pagesAcknowledged);
+
+        return pagesAcknowledged;
+    }
+
+    private synchronized void acknowledgePagesFromMemory(long sequenceId, long startingSequenceId)
+    {
+        checkState(startingSequenceId >= currentMemorySequenceId.get(), "Invalid startingSequenceId for memory pages");
+        int pagesToRemove = toIntExact(sequenceId - startingSequenceId);
+        checkArgument(pagesToRemove <= pages.size(), "Invalid sequenceId");
+
+        long bytesRemoved = 0;
+        for (int i = 0; i < pagesToRemove; i++) {
+            SerializedPage removedPage = pages.remove();
+            bytesRemoved += removedPage.getRetainedSizeInBytes();
+            currentMemorySequenceId.incrementAndGet();
+        }
+
+        totalBufferedPages.addAndGet(-pagesToRemove);
+        totalBufferedBytes.addAndGet(-bytesRemoved);
+        totalPagesRemaining.addAndGet(-pagesToRemove);
+    }
+
+    @Override
+    public void abort(OutputBufferId bufferId)
+    {
+        checkArgument(bufferId.getId() == outputBufferId.getId(), "Invalid outputBufferId");
```

Comment:
Minor nitpick, maybe log the error with the ids mismatched?

Reply:
I don't think this is necessary. The client shouldn't call abort on any other OutputBufferId besides 0. Because we aren't writing to multiple buffers like other OutputBuffer, I made `SPOOLING_OUTPUT_BUFFERS` in OutputBuffers so that we won't be able to add more bufferIds.